### PR TITLE
Connected dApps: a list with empty state, plus button, etc etc / fetching dApps / disconnecting dApps

### LINKED ISSUE #19839: Add WC connected dApps list
# Feature Issue

[comment]: # (Please replace ... with your information. Remove < and >)

### User Story

As a user, I want to know which dApps I am connected to, so that I can manage them in the future (e.g. disconnect)

### Description

The user should see the dApps they're connected to on a specific account. For the time being, we should use the Wallet Connect API for getting the active sessions and render them according to the designs. Shouldn't focus on design fidelity, but rather on the correctness of the flow.

[Design link for dApp list](https://www.figma.com/file/mLyiYPXqu8KetdQONOlZpV/Descoped-dApp-Interactions?type=design&node-id=69-165839&mode=design&t=fh2gA8lqRPPAuWEM-4)

[Design link for dApp button and dApp list empty state](https://www.figma.com/file/mLyiYPXqu8KetdQONOlZpV/Descoped-dApp-Interactions?type=design&node-id=19-134341&mode=design&t=fh2gA8lqRPPAuWEM-4)

### Acceptance Criteria

- [x] The user can see a button for dApps on the wallet account screen
- [x] The user should see the "Connected dApps" screen if the button is pressed
- [x] The user should see the "+" button, which shows a "not implemented" alert
- [ ] The user should see a list of active sessions for the current account, if any
- [x] The user should see the empty state screen if there are no connections

### LINKED ISSUE #19840: Add WC disconnect dApp flow
# Feature Issue

[comment]: # (Please replace ... with your information. Remove < and >)

### User Story

As a user, I want to be able to disconnect from a dApp, so that I will no longer receive requests from it.

### Description

We should have a button on in the connected dApps list, which would disconnect the session for the account. 

[Design link](https://www.figma.com/file/mLyiYPXqu8KetdQONOlZpV/Descoped-dApp-Interactions?type=design&node-id=69-165833&mode=design&t=fh2gA8lqRPPAuWEM-4)

### Acceptance Criteria

- [x] The user can press disconnect on a dApp on the "Connected dApps" screen
- [x] The user is prompted to confirm the disconnection
- [x] The user receives a success toast if the disconnection is successful
- [x] The dApp should no longer be visible in the list if the disconnection was successful

### PR DESCRIPTION
**Pull request by alwx**
Fixes https://github.com/status-im/status-mobile/issues/19839
Fixes https://github.com/status-im/status-mobile/issues/19840

### Summary

https://github.com/status-im/status-mobile/assets/911127/0a80d443-cf08-4512-9d04-50ba3f5a897c

#### Platforms

- Android
- iOS

##### Functional

- wallet connect

status: ready

### CODE REVIEW

**Comment by alwx** [comment]
Now it also includes the disconnect flow and includes fetching of dApps from the WalletConnect API.

https://github.com/status-im/status-mobile/assets/911127/1f60c11d-1efc-4515-89c9-d49496181f5a

File: src/status_im/contexts/chat/messenger/composer/mentions/view.cljs [hunk 1]
```diff
@@ -55,7 +55,7 @@
      [(seq suggestions)])
     [reanimated/view
      {:style (style/container opacity mentions-pos theme)}
-     [rn/flat-list
+     [
```
Comment (clauxx) [inline src/status_im/contexts/chat/messenger/composer/mentions/view.cljs]: removed accidentally i guess

File: src/status_im/contexts/wallet/connected_dapps/disconnect_dapp/view.cljs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    [quo.core :as quo]
+    [react-native.core :as rn]
+    [status-im.contexts.wallet.connected-dapps.disconnect-dapp.style :as style]
+    [utils.i18n :as i18n]
+    [utils.re-frame :as rf]))
+
+(defn view
+  [{:keys [wallet-account dapp on-disconnect]}]
+  (let [{:keys [avatar name]} dapp
+        {:keys [color]}       wallet-account]
```
Comment (clauxx) [inline src/status_im/contexts/wallet/connected_dapps/disconnect_dapp/view.cljs]: to improve performance, might be better to pass only the props it needs (avatar, name, color, on-disconnect) instead of the entire maps

File: src/status_im/contexts/wallet/connected_dapps/view.cljs [hunk 3]
```diff
@@ -1,29 +1,128 @@
...
-      :accessibility-label :close-scan-qr-code
-      :on-press            #(rf/dispatch [:navigate-back])}
-     :i/close]]
-   [quo/text
-    {:size   :heading-1
-     :weight :semi-bold
-     :style  (style/header-text (when subtitle true))}
-    title]])
+  [{:keys [title wallet-account on-close on-add]}]
+  (let [{:keys [color]} wallet-account]
```
Comment (clauxx) [inline src/status_im/contexts/wallet/connected_dapps/view.cljs:59]: same here, if only the color is needed, then we should just pass it instead of passing the whole account map

File: src/status_im/contexts/wallet/connected_dapps/view.cljs [hunk 4]
```diff
@@ -1,29 +1,128 @@
...
+       :on-add         #(js/alert "Feature not implemented")}]
+     (if (empty? pairings)
+       [quo/empty-state
+        {:title           (i18n/label :t/no-dapps)
+         :description     (i18n/label :t/no-dapps-description)
+         :image           (resources/get-themed-image :no-dapps theme)
+         :container-style style/empty-container-style}]
+       [rn/view (style/dapps-container bottom)
+        [rn/flat-list
+         {:data                    pairings
```
Comment (clauxx) [inline src/status_im/contexts/wallet/connected_dapps/view.cljs:102]: I wonder if we should normalize the pairings data in the db, so that we have in the db a list of pairing topics and a map with topics->pairing-data. This way we would only pass the id to the flatlist item, and the item component can retrieve the pairing data with a subscription. Otherwise, we always have to map over the pairings vector to find the relevant one, which might cause perf issues in the future.

File: src/status_im/contexts/wallet/wallet_connect/events.cljs [hunk 5]
```diff
@@ -14,13 +14,19 @@
 (rf/reg-event-fx
  :wallet-connect/on-init-success
  (fn [{:keys [db]} [web3-wallet]]
-   {:db (assoc db :wallet-connect/web3-wallet web3-wallet)
-    :fx [[:dispatch [:wallet-connect/register-event-listeners]]]}))
+   {:db (assoc-in db [:wallet :wallet-connect :web3-wallet] web3-wallet)
```
Comment (clauxx) [inline src/status_im/contexts/wallet/wallet_connect/events.cljs]: I think we should keep the :wallet-connect/... keys as roots and avoid nesting unless necessary. I remember there was a conversation at the offsite about this topic @J-Son89 @cammellos

File: src/status_im/contexts/wallet/common/account_switcher/view.cljs [hunk 6]
```diff
@@ -35,7 +35,9 @@
       :right-side          [(when (and (ff/enabled? ::ff/wallet.wallet-connect)
                                        (not watch-only?))
                               {:icon-name :i/dapps
-                               :on-press  #(rf/dispatch [:navigate-to :screen/wallet.connected-dapps])})
+                               :on-press  (fn []
+                                            (rf/dispatch [:navigate-to
+                                                          :screen/wallet.connected-dapps]))})
```
Comment (clauxx) [inline src/status_im/contexts/wallet/common/account_switcher/view.cljs]: the fn can go outside the component

File: src/status_im/contexts/wallet/wallet_connect/events.cljs (hunk 5 above)
Comment (alwx) [reply src/status_im/contexts/wallet/wallet_connect/events.cljs]: my point here is that it becomes pretty much a mess but interesting to hear other opinions

Comment (clauxx) [reply src/status_im/contexts/wallet/wallet_connect/events.cljs]: Maybe we can have a session to discuss the db structure next week, but for now I'd propose to keep the structure with root keys, otherwise there's going to be some nasty conflicts. WDYT?

File: src/status_im/contexts/wallet/common/account_switcher/view.cljs (hunk 6 above)
Comment (alwx) [reply src/status_im/contexts/wallet/common/account_switcher/view.cljs]: I didn't get what you mean here. 
Why?

File: src/status_im/contexts/wallet/connected_dapps/view.cljs (hunk 3 above)
Comment (alwx) [reply src/status_im/contexts/wallet/connected_dapps/view.cljs:59]: `wallet-account` is being passed to `account-details` component (line 80)

File: src/status_im/contexts/wallet/connected_dapps/disconnect_dapp/view.cljs (hunk 2 above)
Comment (alwx) [reply src/status_im/contexts/wallet/connected_dapps/disconnect_dapp/view.cljs]: Updated it to accept `customization-color` instead of `wallet-account`, but passing `dapp` here makes sense because it's literally a map of just `avatar`, `name` and `url` & it makes sense to be passed here also for indicating that `disconnect-dapp` component requires `dapp` to be passed.

File: src/status_im/contexts/wallet/connected_dapps/view.cljs (hunk 4 above)
Comment (alwx) [reply src/status_im/contexts/wallet/connected_dapps/view.cljs:102]: But we also want to maintain the order, right? I'd argue that the "display list" operation will be used more often, and, in addition to that, even finding something in a list of 1000 items will take about zero time compared to the amount it will take to redraw the list in the case of update/delete opration anyway.

Comment (alwx) [reply src/status_im/contexts/wallet/connected_dapps/view.cljs:102]: But also I think it doesn't matter. We're here making a list that will contain zero to five items for the majority of users.

File: src/status_im/contexts/wallet/common/account_switcher/view.cljs (hunk 6 above)
Comment (clauxx) [reply src/status_im/contexts/wallet/common/account_switcher/view.cljs]: so the function is not recreated on each render

File: src/status_im/contexts/wallet/wallet_connect/events.cljs (hunk 5 above)
Comment (clauxx) [reply src/status_im/contexts/wallet/wallet_connect/events.cljs]: i'd say it's best to change it to how it was to avoid unnecessary conflicts, since @shivekkhurana and me are already working with the structure we have on develop atm, but we can discuss and agree on a way during today's session 👍

File: src/quo/components/list_items/dapp/view.cljs [hunk 7]
```diff
@@ -1,22 +1,22 @@
...
     [quo.components.list-items.dapp.style :as style]
     [quo.components.markdown.text :as text]
-    [quo.foundations.colors :as colors]
     [quo.theme :as quo.theme]
     [react-native.core :as rn]
     [react-native.fast-image :as fast-image]))
 
 (defn view
-  [{:keys [dapp action on-press on-press-icon] :as props}]
+  [{:keys [dapp on-press right-component] :as props}]
```
Comment (J-Son89) [inline src/quo/components/list_items/dapp/view.cljs]: why are we passing in a component here? the design system only shows one variant of button? 
Does it need to reflect the alternatives better? 
Is the right hand side actually generic? 🤔

Comment (J-Son89) [reply src/quo/components/list_items/dapp/view.cljs]: imo it's better to encapsulate all of the styles in the design system that we activate via configuration then handling with dynamic components unless really needed 👍

File: src/status_im/contexts/wallet/wallet_connect/events.cljs (hunk 5 above)
Comment (alwx) [reply src/status_im/contexts/wallet/wallet_connect/events.cljs]: It's changed now

File: src/quo/components/list_items/dapp/view.cljs [hunk 8]
```diff
@@ -1,22 +1,22 @@
...
   (let [theme                  (quo.theme/use-theme)
         [pressed? set-pressed] (rn/use-state false)
         on-press-in            (rn/use-callback #(set-pressed true))
         on-press-out           (rn/use-callback #(set-pressed false))]
     [rn/pressable
      {:style        (style/container (assoc props :pressed? pressed?))
-      :on-press     on-press
+      :testID       "dapp-pressable"
```
Comment (J-Son89) [inline src/quo/components/list_items/dapp/view.cljs]: we should avoid testID's unless really needed, accessibility labels much better.
https://testing-library.com/docs/queries/about/

File: src/quo/components/navigation/page_nav/view.cljs [hunk 9]
```diff
@@ -30,16 +30,17 @@
...
-            {:type                (button-type background)
-             :icon-only?          true
-             :size                32
-             :on-press            on-press
-             :background          (if behind-overlay?
-                                    :blur
-                                    (when (button-properties/backgrounds background) background))
-             :accessibility-label accessibility-label}
-            icon-name])]
+           [rn/view {:flex-grow 1 :flex-basis 1}
```
Comment (J-Son89) [inline src/quo/components/navigation/page_nav/view.cljs]: can you put this in a `:style { }` block?

Comment (J-Son89) [reply src/quo/components/navigation/page_nav/view.cljs]: also there should be `container-style` prop for this purpose. I think the button has it 🤔

File: src/status_im/contexts/wallet/connected_dapps/view.cljs [hunk 10]
```diff
@@ -1,29 +1,128 @@
 (ns status-im.contexts.wallet.connected-dapps.view
   (:require
     [quo.core :as quo]
+    [quo.foundations.colors :as colors]
+    [quo.theme :as quo.theme]
```
Comment (J-Son89) [inline src/status_im/contexts/wallet/connected_dapps/view.cljs]: nit [quo.theme]

File: src/quo/components/list_items/dapp/view.cljs (hunk 7 above)
Comment (alwx) [reply src/quo/components/list_items/dapp/view.cljs]: This dapp component can also be used in wallet, there is a `dapps` tab here where it comes without anything. In addition, there is also a variation of it with "..." icon instead of disconnect button.

File: src/quo/components/navigation/page_nav/view.cljs (hunk 9 above)
Comment (alwx) [reply src/quo/components/navigation/page_nav/view.cljs]: this one is used to align icon within the `container` — adding this code to either container or the button itself leads to completely different results

but moved this code to styles

File: src/status_im/contexts/chat/messenger/composer/mentions/view.cljs [hunk 11]
```diff
@@ -55,7 +55,7 @@
      [(seq suggestions)])
     [reanimated/view
      {:style (style/container opacity mentions-pos theme)}
-     [rn/flat-list
+     [rn/view
```
Comment (clauxx) [inline src/status_im/contexts/chat/messenger/composer/mentions/view.cljs]: do we need this change in the composer? seems unrelated

**Review by clauxx — APPROVED** [review]

File: src/status_im/contexts/chat/messenger/composer/mentions/view.cljs (hunk 11 above)
Comment (alwx) [reply src/status_im/contexts/chat/messenger/composer/mentions/view.cljs]: damn, that's a stupid one, removing

**Comment by clauxx** [comment]
Looks good, but I think we'll need QA manual check for the bits that were changed outside Wallet Connect. WDYT @alwx?

**Comment by alwx** [comment]
@clauxx I think the only of those bits is really the dapp item in Wallet but the thing is it's not even used now.
So the only thing that can be checked here is `page-nav` title again.

**Comment by alwx** [comment]
@status-im/mobile-qa  One thing to check is if the `page-nav` title is fine (the same check was done here: https://github.com/status-im/status-mobile/pull/19934)

File: src/status_im/contexts/wallet/wallet_connect/events.cljs [hunk 12]
```diff
@@ -53,6 +82,15 @@
              :on-fail     #(log/error "Failed to pair with dApp" {:error %})
              :on-success  #(log/info "dApp paired successfully")}]]})))
 
+(rf/reg-event-fx
+ :wallet-connect/fetch-active-sessions
+ (fn [{:keys [db]}]
+   (let [web3-wallet (get db :wallet-connect/web3-wallet)]
+     {:fx [[:effects.wallet-connect/fetch-active-sessions
+            {:web3-wallet web3-wallet
+             :on-fail     #(log/error "Failed to get active sessions" {:error %})
```
Comment (shivekkhurana) [inline src/status_im/contexts/wallet/wallet_connect/events.cljs:91]: `on-fail` can be `on-error` for consistency: https://github.com/status-im/status-mobile/pull/20199#discussion_r1617772934

File: src/status_im/contexts/wallet/connected_dapps/view.cljs [hunk 13]
```diff
@@ -1,29 +1,129 @@
...
+  (let [{:keys [bottom]}                   (safe-area/get-insets)
+        {:keys [color] :as wallet-account} (rf/sub [:wallet/current-viewing-account])
+        pairings                           (rf/sub [:wallet-connect/pairings])
+        theme                              (quo.theme/use-theme)]
+    [rn/view {:flex 1}
+     [header
+      {:title          (i18n/label :t/connected-dapps)
+       :wallet-account wallet-account
+       :on-close       #(rf/dispatch [:navigate-back])
+       :on-add         #(js/alert "Feature not implemented")}]
```
Comment (shivekkhurana) [inline src/status_im/contexts/wallet/connected_dapps/view.cljs:93]: There is a `not-implemented` namespace with this alert as a function. 

It's better to use that so in future we can do a global lookup of functions that need to be implemented.

File: src/status_im/contexts/wallet/connected_dapps/view.cljs (hunk 3 above)
Comment (shivekkhurana) [reply src/status_im/contexts/wallet/connected_dapps/view.cljs:59]: Not a requirement, but the Clojure can eliminate this `let` bind:

```clj
(defn- header
  [{:keys [title on-close on-add] 
    :as args 
    {:keys [color]} :wallet-account}]
    ...)
```

**Review by shivekkhurana — APPROVED** [review]

**Comment by VolodLytvynenko** [comment]
Hi @alwx, thank you for the PR. No PR related issues from my. There's one issue related to enabling the connection dApps feature flag, but it's also present in the last nightly build. So, It can be fixed separately. Please let me know if it will be fixed within the current PR. Otherwise, the PR can be merged

## ISSUE 1: Error "cannot read property 'init' of undefined" occurs after relogin when 'wallet connect' feature flag enabled

### Steps:
1. Enable 'wallet connect' feature flag
2. Reloging

### Actual result:
image

### OS: 
IOS, Android

### Devices:
- Pixel 7a, Android 13
- iPhone 11 Pro Max, IOS 17

### Logs
[Status-debug-logs.zip](https://github.com/status-im/status-mobile/files/15483679/Status-debug-logs.zip)

**Comment by clauxx** [comment]
@VolodLytvynenko Yes, the wallet connect functionality is not yet ready to be tested. We need to add the project-id secret so it's available in the builds, so this error is expected.

**Comment by VolodLytvynenko** [comment]
> @VolodLytvynenko Yes, the wallet connect functionality is not yet ready to be tested. We need to add the project-id secret so it's available in the builds, so this error is expected.

thank you for clarification. PR can be merged. No other issues from my side

File: src/status_im/contexts/wallet/connected_dapps/view.cljs (hunk 13 above)
Comment (clauxx) [reply src/status_im/contexts/wallet/connected_dapps/view.cljs:93]: You're right, for features that won't be implemented for a while we must use the same function for easier discoverability. Here though it won't stay like this for long, as `on-add` will be implemented in the coming weeks anyways and without needing a lookup, so should be fine for now. WDYT?
